openvpn/init.sls

```diff
@@ -38,7 +38,7 @@ openvpn_external_repo:
   {%- for dh in map.dh_files %}
 openvpn_create_dh_{{ dh }}:
   cmd.run:
-    - name: openssl dhparam -out {{ map.conf_dir }}/dh{{ dh }}.pem {{ dh }}
+    - name: openssl dhparam -dsaparam -out {{ map.conf_dir }}/dh{{ dh }}.pem {{ dh }}
```

Comment:
I have read that StackExchange answer before. It seems like a poor justification for making this change without further evidence, no?

Reply:
I am not a security expert by any means.

But for example, running `openssl dhparam -out 4096.pem 4096` without `-dsaparam` can take a few hours, which imo is not at all acceptable inside a Salt state.

The cleanest solution would probably be to make dhparam creation optional, so you could either:
* Create dhparams without dsaparam
* Create dhparams with dsaparam
* Don't Create dhparam at all (user provides his own)

Reply:
It is currently possible to override the defaults to only create ```2048``` dhparams, which I believe takes only a few minutes.

If you wanted to make further changes that were optional/configurable, I wouldn't have a problem with that.

Reply:
> can take a few hours, which imo is not at all acceptable inside a Salt state.

That depends. But usually faster is better. :-)

> Create dhparams with/out dsaparam

https://www.wakusi.de/forum/viewtopic.php?t=337 (German; sry, didn't find an equivalent english source)
https://en.wikipedia.org/wiki/Strong_prime
https://security.stackexchange.com/questions/42415/openvpn-dhparam (last two paragraphs)

I'm not a cryptographer, but I've got a fairly strong security background. (Which is completely irrelevant to the argument. I could still be wrong.)

As far as I understand it `dsaparam` is permissible when _each and every_ VPN peer generates its own DH parameters. This formula does just that. :-) (At least if one equates the _"for each use"_ from http://openssl.cs.utah.edu/docs/apps/dhparam.html with _"for each peer"_.)
Furthermore many people (with cryptographical background) seem to consider the security of strong primes vs large primes roughly equal.

 That being said, I propose we provide a fast, secure default along with the option to drop `dsaparam`:

- default (as implemented by @marconett ): `openssh dhparam -dsaparam -out 4096.pem 4096`
- opt-out for `-dsaparam` in Pillar: `dsaparam: False` (comment: `# use strong primes`)

@marconett @dafyddj Would that be OK with you?

Personally, I'm going to use "safe/strong" primes because the Logjam mitigation docs tell me to: https://weakdh.org/sysadmin.html See also: https://en.wikipedia.org/wiki/Logjam_(computer_security)

> Don't Create dhparam at all (user provides his own)

If the file is present, it won't be overwritten.

(I oppose sharing DH parameters across machines or even services. One vulnerability in a service could leak the DH parameters of an entire infrastructure.)

Reply:
Thanks for the effort you put in looking this up. I am in favor of making it optional using pillar data.

Reply:
My reading of [this security announcement](https://www.openssl.org/news/secadv/20160128.txt) suggests that *for each use* literally means for each handshake. 
Additionally it says ```It is possible to generate parameters using non "safe" primes,
but this option has always been documented as requiring single use and is not
the default or believed to be common.``` 
So I struggle to agree to make something considered unusual or not common the default in the formula. Configurable, yes; default, no.

Reply:
@dafyddj Thank you very much, I just missed that advisory. (May I ask: How did you find it?)

The advisory mentions `SSL_OP_SINGLE_DH_USE`.

>  It is believed that many popular applications do set this option and would therefore not be at risk.

OpenSSL (and therefore OpenVPN) has already been patched:

> OpenSSL _before_ 1.0.2f will reuse the key if […]
> It will not reuse the key for DHE ciphers suites if […] SSL_OP_SINGLE_DH_USE is set
> […]
> The fix for this issue adds an additional check where a "q" parameter is
available (as is the case in X9.42 based parameters). This detects the
only known attack, and is the only possible defense for static DH ciphersuites.
This could have some performance impact.
>
>**Additionally the SSL_OP_SINGLE_DH_USE option has been switched on by default
and cannot be disabled.** This could have some performance impact.

(Emphasis by me.)

So I still think one can use `-dsaparam` securely, **if** OpenSSL >= 1.0.2f is used, but for me personally: I won't take the risk (of being wrong) either. :-)

This is the first limit for using `-dsaparam`. There is another (which I missed in my last post, sry):

> This [`-dsaparam`] is supported since the 1.0.2
  version. Older versions can't read files generated in this way.

If we set `-dsaparam` as a default, the formula won't work for (hopefully still maintained) older versions of OpenVPN. Therefore it can only become an opt-in.

@marconett If the size of the DH parameter is set to 2048 (instead of 4096), the generation takes way less time. This is already configurable via Pillar.
Since your stated goal is fast(er) DH parameter generation, I guess our discussion boils down to:

- Do we switch back to a default of 2048 bits? (Still large enough, but faster!)
- Do we still want `-dsaparam` as an opt-in in this formula?

@marconett @dafyddj I answer these 'yes' and 'yes'. How about you?

Reply:
@alxwr the security advisory was linked at the very bottom of the StackExchange question you referred to.

> Do we switch back to a default of 2048 bits? (Still large enough, but faster!)

That's an acceptable compromise, I think.

> Do we still want `-dsaparam` as an opt-in in this formula?

It wouldn't hurt, other than the trouble of coding it up.

Reply:
@dafyddj memo to myself: always read stuff top to bottom. :-) Thanks!